### PR TITLE
Add woox pro

### PR DESCRIPTION
**Pull request by sc0Vu**
There is websocket api for woo exchange: https://docs.woo.org/#websocket-api-v2.

In this PR, I integrate woo into ccxt.pro. They also support positioninfo topic, do we have specification of `watchPositions`/`watchPosition`?

### CODE REVIEW

**Comment by frosty00** [comment]
> In this PR, I integrate woo into ccxt.pro. They also support positioninfo topic, do we have specification of watchPositions/watchPosition?

We currently don't have a spec for this yet. I will review your bybit PR tomorrow and then will take a look at this.

File: js/pro/woo.js [hunk 1]
```diff
@@ -0,0 +1,722 @@
...
+                },
+                'test': {
+                    'ws': {
+                        'public': 'wss://wss.staging.woo.org/ws/stream',
+                        'private': 'wss://wss.staging.woo.org/v2/ws/private/stream',
+                    },
+                },
+            },
+            'requiredCredentials': {
+                'apiKey': false,
```
Comment (carlosmiei) [inline js/pro/woo.js:41]: why false here?

File: js/pro/woo.js [hunk 2]
```diff
@@ -0,0 +1,722 @@
...
+            },
+            'streaming': {
+                'ping': this.ping,
+                'keepAlive': 10000,
+            },
+        });
+    }
+
+    requestId (url) {
+        const options = this.safeValue (this.options, 'requestId', {});
```
Comment (carlosmiei) [inline js/pro/woo.js:58]: if the connection breaks and the user re-connects, shouldn't we reset this counter?

Example:
```Javacript
    while(true) {
        try{
            trades = await exchange.watchTrades('BTC/USDT');
        } catch(e) {
            // do nothing
        }
    }
```

File: js/pro/woo.js [hunk 3]
```diff
@@ -0,0 +1,722 @@
...
+        }
+        const timestamp = this.safeInteger (message, 'ts');
+        const snapshot = this.parseOrderBook (data, symbol, timestamp, 'bids', 'asks');
+        orderbook.reset (snapshot);
+        client.resolve (orderbook, topic);
+    }
+
+    async watchTicker (symbol, params = {}) {
+        await this.loadMarkets ();
+        const name = 'ticker';
```
Comment (carlosmiei) [inline js/pro/woo.js:129]: @sc0Vu mainnet seems down,  is just me?
image

File: js/pro/woo.js [hunk 4]
```diff
@@ -0,0 +1,722 @@
...
+            'vwap': undefined,
+            'open': this.safeString (ticker, 'open'),
+            'close': this.safeString (ticker, 'close'),
+            'last': undefined,
+            'previousClose': undefined,
+            'change': undefined,
+            'percentage': undefined,
+            'average': undefined,
+            'baseVolume': this.safeString (ticker, 'volume'),
+            'quoteVolume': undefined,
```
Comment (carlosmiei) [inline js/pro/woo.js:173]: isn't the amount the quote volume? 
image

File: js/pro/woo.js [hunk 5]
```diff
@@ -0,0 +1,722 @@
...
+        //     }
+        // }
+        //
+        const topic = this.safeString (message, 'topic');
+        const timestamp = this.safeInteger (message, 'ts');
+        const data = this.safeValue (message, 'data');
+        const marketId = this.safeString (data, 'symbol');
+        const market = this.safeMarket (marketId);
+        const symbol = market['symbol'];
+        const trade = this.parseWsTrade (this.extend (data, { timestamp }), market);
```
Comment (carlosmiei) [inline js/pro/woo.js:367]: This is not working in python:

```
  File "/opt/homebrew/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
  File "/Users/cjg/Git/ccxt10/ccxt/python/ccxt/pro/base/fast_client.py", line 25, in handler
    self.handle_message(message)
  File "/Users/cjg/Git/ccxt10/ccxt/python/ccxt/pro/base/aiohttp_client.py", line 32, in handle_message
    self.handle_text_or_binary_message(message.data)
  File "/Users/cjg/Git/ccxt10/ccxt/python/ccxt/pro/base/aiohttp_client.py", line 27, in handle_text_or_binary_message
    self.on_message_callback(self, decoded)
  File "/Users/cjg/Git/ccxt10/ccxt/python/ccxt/pro/woo.py", line 616, in handle_message
    return method(client, message)
  File "/Users/cjg/Git/ccxt10/ccxt/python/ccxt/pro/woo.py", line 352, in handle_trade
    trade = self.parse_ws_trade(self.extend(data, {timestamp}), market)
  File "/Users/cjg/Git/ccxt10/ccxt/python/ccxt/base/exchange.py", line 927, in extend
    result.update(arg)
TypeError: cannot convert dictionary update sequence element #0 to a sequence
```

File: js/pro/woo.js [hunk 6]
```diff
@@ -0,0 +1,722 @@
...
+        const requestId = this.requestId (url);
+        const subscribe = {
+            'id': requestId,
+        };
+        const request = this.extend (subscribe, message);
+        return await this.watch (url, messageHash, request, messageHash, subscribe);
+    }
+
+    async watchOrders (symbol = undefined, since = undefined, limit = undefined, params = {}) {
+        await this.loadMarkets ();
```
Comment (carlosmiei) [inline js/pro/woo.js:463]: I'm also getting this signature error:
image

File: js/pro/woo.js (hunk 2 above)
Comment (frosty00) [reply js/pro/woo.js:58]: actually the requestId doesn't matter, just needs to be unique afaik

File: js/pro/woo.js (hunk 1 above)
Comment (sc0Vu) [reply js/pro/woo.js:41]: should be true...

File: js/pro/woo.js (hunk 3 above)
Comment (sc0Vu) [reply js/pro/woo.js:129]: Not sure about this, but test net works for me.

File: js/pro/woo.js (hunk 6 above)
Comment (sc0Vu) [reply js/pro/woo.js:463]: It worked for me.

$ node examples/js/cli woo watchOrders --test --verbose
Node.js: v18.8.0
CCXT v2.1.96
woo.watchOrders ()
2022-12-05T06:40:51.243Z connecting to wss://wss.staging.woo.org/v2/ws/private/stream/...
2022-12-05T06:40:51.335Z onUpgrade
2022-12-05T06:40:51.336Z onOpen
2022-12-05T06:40:51.336Z sending {
  event: 'auth',
  params: {
    apikey: '...',
    sign: '...',
    timestamp: '1670222451242'
  }
}
2022-12-05T06:40:51.392Z onMessage { event: 'auth', success: true, ts: 1670222451420 }
2022-12-05T06:40:51.492Z sending { id: 1, event: 'subscribe', topic: 'executionreport' }
2022-12-05T06:40:51.545Z onMessage { id: '1', event: 'subscribe', success: true, ts: 1670222451577 }
2022-12-05T06:40:59.976Z onMessage { event: 'ping', ts: 1670222460000 }
2022-12-05T06:41:01.337Z sending { event: 'ping' }
2022-12-05T06:41:01.391Z onMessage { event: 'pong', ts: 1670222461419 }

Comment (carlosmiei) [reply js/pro/woo.js:463]: @sc0Vu humm weird, will try again

File: js/pro/woo.js [hunk 7]
```diff
@@ -0,0 +1,722 @@
...
+        const name = 'orderbook';
+        const market = this.market (symbol);
+        const topic = market['id'] + '@' + name;
+        const request = {
+            'event': 'subscribe',
+            'topic': topic,
+        };
+        const message = this.extend (request, params);
+        const orderbook = await this.watchPublic (topic, message);
+        return orderbook.limit (limit);
```
Comment (carlosmiei) [inline js/pro/woo.js:87]: @sc0Vu `limit` does not take any argument

**Comment by carlosmiei** [comment]
LGTM ✅ 

```
p woo watchTicker "BTC/USDT" --sandbox
Python v3.9.7
CCXT v2.2.68
woo.watchTicker(BTC/USDT)
{'ask': None,
 'askVolume': None,
 'average': 17085.43,
 'baseVolume': 587.39941959,
 'bid': None,
 'bidVolume': None,
 'change': 245.14,
 'close': 17208.0,
 'datetime': '2022-12-05T14:41:29.000Z',
 'high': 17403.89,
 'info': {'aggregatedAmount': 3989304763.9109926,
          'aggregatedQuantity': 231835.31830062,
          'amount': 10159316.79138555,
          'close': 17208.0,
          'count': 1057,
          'date': 1670251289000,
          'high': 17403.89,
          'lastTs': 1670251288751,
          'low': 16955.5,
          'open': 16962.86,
          'symbol': 'SPOT_BTC_USDT',
          'volume': 587.39941959},
 'last': 17208.0,
 'low': 16955.5,
 'open': 16962.86,
 'percentage': 1.4451572435308668,
 'previousClose': None,
 'quoteVolume': 10159316.79138555,
 'symbol': 'BTC/USDT',
 'timestamp': 1670251289000,
 'vwap': 17295.415100131817}
{'ask': None,
 'askVolume': None,
 'average': 17084.83,
 'baseVolume': 587.39941959,
 'bid': None,
 'bidVolume': None,
 'change': 243.86,
 'close': 17206.76,
 'datetime': '2022-12-05T14:41:30.000Z',
 'high': 17403.89,
 'info': {'aggregatedAmount': 3989240972.395783,
          'aggregatedQuantity': 231831.55431294,
          'amount': 10159316.79138555,
          'close': 17206.76,
          'count': 1057,
          'date': 1670251290000,
          'high': 17403.89,
          'lastTs': 1670251289951,
          'low': 16955.5,
          'open': 16962.9,
          'symbol': 'SPOT_BTC_USDT',
          'volume': 587.39941959},
 'last': 17206.76,
 'low': 16955.5,
 'open': 16962.9,
 'percentage': 1.437607956186737,
 'previousClose': None,
 'quoteVolume': 10159316.79138555,
 'symbol': 'BTC/USDT',
 'timestamp': 1670251290000,
 'vwap': 17295.415100131817}
 ```

```
p woo watchTicker "BTC/USDT" --sandbox
Python v3.9.7
CCXT v2.2.68
woo.watchTicker(BTC/USDT)
{'ask': None,
 'askVolume': None,
 'average': 17085.43,
 'baseVolume': 587.39941959,
 'bid': None,
 'bidVolume': None,
 'change': 245.14,
 'close': 17208.0,
 'datetime': '2022-12-05T14:41:29.000Z',
 'high': 17403.89,
 'info': {'aggregatedAmount': 3989304763.9109926,
          'aggregatedQuantity': 231835.31830062,
          'amount': 10159316.79138555,
          'close': 17208.0,
          'count': 1057,
          'date': 1670251289000,
          'high': 17403.89,
          'lastTs': 1670251288751,
          'low': 16955.5,
          'open': 16962.86,
          'symbol': 'SPOT_BTC_USDT',
          'volume': 587.39941959},
 'last': 17208.0,
 'low': 16955.5,
 'open': 16962.86,
 'percentage': 1.4451572435308668,
 'previousClose': None,
 'quoteVolume': 10159316.79138555,
 'symbol': 'BTC/USDT',
 'timestamp': 1670251289000,
 'vwap': 17295.415100131817}
 ```

 ```
 
(base) ➜  ccxt git:(add-woo-pro) ✗ p woo watchOHLCV "BTC/USDT" --sandbox
Python v3.9.7
CCXT v2.2.68
woo.watchOHLCV(BTC/USDT)
[[1670251440000, 17211.4, 17213.0, 17208.38, 17208.6, 22.32458309]]
[[1670251440000, 17211.4, 17213.0, 17208.38, 17208.6, 23.11464309]]
[[1670251440000, 17211.4, 17213.0, 17208.38, 17209.14, 23.79047309]]
[[1670251440000, 17211.4, 17213.0, 17208.38, 17208.99, 24.06878853]]
```

```
✗ p woo watchTickers '["BTC/USDT", "LTC/USDT"]' --sandbox
Python v3.9.7
CCXT v2.2.68
woo.watchTickers(['BTC/USDT', 'LTC/USDT'])
{'BTC/USDT': {'ask': None,
              'askVolume': None,
              'average': 17097.95,
              'baseVolume': 580.17310159,
              'bid': None,
              'bidVolume': None,
              'change': 247.98,
              'close': 17221.94,
              'datetime': '2022-12-05T14:49:31.000Z',
              'high': 17403.89,
              'info': {'aggregatedAmount': 3970665597.2280755,
                       'aggregatedQuantity': 230720.51653792,
                       'amount': 10036553.18869695,
                       'close': 17221.94,
                       'count': 1047,
                       'date': 1670251771000,
                       'high': 17403.89,
                       'lastTs': 1670251769351,
                       'low': 16967.57,
                       'open': 16973.96,
                       'symbol': 'SPOT_BTC_USDT',
                       'volume': 580.17310159},
              'last': 17221.94,
              'low': 16967.57,
              'open': 16973.96,
              'percentage': 1.4609437043565554,
              'previousClose': None,
              'quoteVolume': 10036553.18869695,
              'symbol': 'BTC/USDT',
              'timestamp': 1670251771000,
              'vwap': 17299.23907397837},
 'LTC/USDT': {'ask': None,
              'askVolume': None,
              'average': 79.745,
              'baseVolume': 397.340832,
              'bid': None,
              'bidVolume': None,
              'change': 7.17,
              'close': 83.33,
              'datetime': '2022-12-05T14:49:31.000Z',
              'high': 84.98,
              'info': {'aggregatedAmount': 136170286.66076213,
                       'aggregatedQuantity': 1672129.990059,
                       'amount': 32941.93380432,
                       'close': 83.33,
                       'count': 11,
                       'date': 1670251771000,
                       'high': 84.98,
                       'lastTs': 1670251769615,
                       'low': 75.91,
                       'open': 76.16,
                       'symbol': 'SPOT_LTC_USDT',
                       'volume': 397.340832},
              'last': 83.33,
              'low': 75.91,
              'open': 76.16,
              'percentage': 9.41439075630252,
              'previousClose': None,
              'quoteVolume': 32941.93380432,
              'symbol': 'LTC/USDT',
              'timestamp': 1670251771000,
              'vwap': 82.90598687909326}}
```

```
 p woo watchOrders --sandbox
Python v3.9.7
CCXT v2.2.68
woo.watchOrders()
[{'amount': 0.001,
  'average': None,
  'clientOrderId': '0',
  'cost': '0.0',
  'datetime': '2022-12-05T14:51:54.545Z',
  'fee': {'cost': '0.0', 'currency': 'BTC'},
  'filled': 0.0,
  'id': '53559747',
  'info': {'clientOrderId': 0,
           'executedPrice': 0.0,
           'executedQuantity': 0.0,
           'fee': 0.0,
           'feeAsset': 'BTC',
           'maker': False,
           'orderId': 53559747,
           'orderTag': 'default',
           'price': 15000.0,
           'quantity': 0.001,
           'reason': '',
           'reduceOnly': False,
           'side': 'BUY',
           'status': 'NEW',
           'symbol': 'SPOT_BTC_USDT',
           'timestamp': 1670251914545,
           'totalExecutedQuantity': 0.0,
           'totalFee': 0.0,
           'tradeId': 0,
           'type': 'LIMIT',
           'visible': 0.001},
  'lastTradeTimestamp': 1670251914545,
  'postOnly': None,
  'price': 15000.0,
  'remaining': 0.001,
  'side': 'buy',
  'status': 'open',
  'stopPrice': None,
  'symbol': 'BTC/USDT',
  'timeInForce': None,
  'timestamp': 1670251914545,
  'trades': None,
  'type': 'limit'}]
  ```

**Review by carlosmiei — APPROVED** [review]

**Pull request merged**